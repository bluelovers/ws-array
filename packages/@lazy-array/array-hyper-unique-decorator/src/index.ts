import { array_unique, array_unique_overwrite, IOptions } from 'array-hyper-unique';

function _core<T extends any = any>(fn: typeof array_unique, options?: IOptions<T>)
{
	return <T extends Function>(target: object,
		propertyKey: string | symbol,
		descriptor: TypedPropertyDescriptor<T>
	) =>
	{
		const old = descriptor.value;

		// @ts-ignore
		descriptor.value = function (...argv)
		{
			return fn(old.apply(this, argv), options)
		}
	}
}

function _coreAsync<T extends any = any>(fn: typeof array_unique, options?: IOptions<T>)
{
	return <T extends () => Promise<any>>(target: object,
		propertyKey: string | symbol,
		descriptor: TypedPropertyDescriptor<T>
	) =>
	{
		const old = descriptor.value;

		// @ts-ignore
		descriptor.value = function (...argv)
		{
			return old.apply(this, argv)
				.then(value => fn(value, options))
		}
	}
}

export function ArrayUniqueDecorator<T extends any = any>(options?: IOptions<T>)
{
	return _core(array_unique, options)
}

export function ArrayUniqueOverwriteDecorator<T extends any = any>(options?: IOptions<T>)
{
	return _core(array_unique_overwrite, options)
}

export function ArrayUniqueAsyncDecorator<T extends any = any>(options?: IOptions<T>)
{
	return _coreAsync(array_unique, options)
}

export function ArrayUniqueOverwriteAsyncDecorator<T extends any = any>(options?: IOptions<T>)
{
	return _coreAsync(array_unique_overwrite, options)
}

export default ArrayUniqueDecorator
