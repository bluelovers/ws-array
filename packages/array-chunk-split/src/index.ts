import { ITSRequireAtLeastOne } from 'ts-type/lib/type/record';

/**
 * same as T[][]
 */
export interface IChunkArray<T> extends Array<T[]>
{

}

/**
 * will calls a defined callback function on each element of an array, and returns an array that contains the results.
 */
export interface IMapCallback<T, R = T>
{
	(value: T[], index: number, array: IChunkArray<T>): R

	(value: T[], index: number, array: IChunkArray<T>): T
}

export type IOptions<T> = {

	/**
	 * source array
	 */
	inputArray: T[],

} & ITSRequireAtLeastOne<{

	/**
	 * Split an array into arrays with max chunk length
	 */
	maxChunkLength?: number,
	/**
	 * Split an array into arrays of chunk with max size
	 */
	maxChunkSize?: number,

}, 'maxChunkLength' | 'maxChunkSize'>;

/**
 * by default will return a array with first value in chunk
 *
 * if mapMethod = true will return last value of chunk
 *
 * if give mapMethod is function
 * will calls a defined callback function on each element of an array, and returns an array that contains the results.
 */
export function arrayChunkMap<T, R = T>(options: IOptions<T> & {
	mapMethod: IMapCallback<T, R>
}): R[]
/**
 * by default will return a array with first value in chunk
 *
 * if mapMethod = true will return last value of chunk
 *
 * if give mapMethod is function
 * will calls a defined callback function on each element of an array, and returns an array that contains the results.
 */
export function arrayChunkMap<T>(options: IOptions<T> & {
	mapMethod: boolean
}): T[]
/**
 * by default will return a array with first value in chunk
 *
 * if mapMethod = true will return last value of chunk
 *
 * if give mapMethod is function
 * will calls a defined callback function on each element of an array, and returns an array that contains the results.
 */
export function arrayChunkMap<T, R = T>(options: IOptions<T> & {
	mapMethod?: boolean | IMapCallback<T, R>
}): R[]
export function arrayChunkMap<T, R = T>(options: IOptions<T> & {
	mapMethod?: boolean | IMapCallback<T, R>
}): R[]
{
	const { inputArray, maxChunkLength, maxChunkSize } = options;
	let { mapMethod } = options;

	let result: IChunkArray<T>;

	if (maxChunkLength != null)
	{
		result = arrayChunkSplit(inputArray, maxChunkLength);
	}
	else if (maxChunkSize != null)
	{
		result = arrayChunkBySize(inputArray, maxChunkSize);
	}
	else
	{
		throw new TypeError(`maxChunkLength or maxChunkSize is required`)
	}

	if (typeof mapMethod !== 'function')
	{
		if (mapMethod)
		{
			// @ts-ignore
			mapMethod = (value) => value[value.length - 1];
		}
		else
		{
			// @ts-ignore
			mapMethod = (value) => value[0];
		}
	}

	return result.map(mapMethod as any) as any as R[];
}

/**
 * Split an array into arrays of chunk with max size
 *
 * @example arrayChunkBySize([1, 2, 3, 4, 5, 6, 7, 8], 5); // => [[1, 2, 3, 4, 5], [6, 7, 8]]
 */
export function arrayChunkBySize<T>(arr: T[], maxChunkSize: number): IChunkArray<T>
{
	const result: IChunkArray<T> = [];
	//let part: T[] = [];

	const { length } = arr;

	if (typeof maxChunkSize !== 'number' || maxChunkSize < 1)
	{
		throw new RangeError(`expected maxChunkSize > 0 but got ${maxChunkSize}`)
	}

	for (let i = 0; i < length; i++)
	{
		let next = i + maxChunkSize;

		result.push(arr.slice(i, next));

		i = next - 1;
	}

	return result;
}

/**
 * Split an array into arrays with max chunk length
 *
 * @example arrayChunkSplit([1, 2, 3, 4, 5, 6, 7, 8], 4); // => [[1, 2], [3, 4], [5, 6], [7, 8]]
 */
export function arrayChunkSplit<T>(arr: T[], maxChunkLength: number)
{
	if (typeof maxChunkLength !== 'number' || maxChunkLength < 1)
	{
		throw new RangeError(`expected maxChunkLength > 0 but got ${maxChunkLength}`)
	}

	const maxChunkSize = Math.max(Math.round(arr.length / maxChunkLength), 1);
	return arrayChunkBySize(arr, maxChunkSize);
}

export default arrayChunkSplit
