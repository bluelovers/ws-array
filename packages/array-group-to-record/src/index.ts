import hashSum from 'hash-sum';

export interface IOptionsForRecord<T>
{
	getKey?(item: T, index: number, arr: T[]): string

	init?(): Record<string, T[]>,
}

export interface IOptionsForMap<T>
{
	getKey?(item: T, index: number, arr: T[]): any

	init?(): Map<any, T[]>,
}

export function handleOptions<O extends IOptionsForRecord<any> | IOptionsForMap<any>>(options: O): O
{
	const getKey = options?.getKey ?? ((item) => hashSum(item));

	return {
		...options,
		getKey,
	}
}

export function arrayGroupToRecord<T>(arr: T[], options?: IOptionsForRecord<T>)
{
	const { getKey, init } = handleOptions(options);

	return arr.reduce((map, item, index, arr) =>
	{

		const id = getKey(item, index, arr);

		map[id] ??= [];
		map[id].push(item);

		return map
	}, init?.() ?? {} as Record<string, T[]>);
}

export function arrayGroupToMap<T>(arr: T[], options?: IOptionsForMap<T>)
{
	const { getKey, init } = handleOptions(options);

	return arr.reduce((map, item, index, arr) =>
	{

		const id = getKey(item, index, arr);

		const a = map.get(id) ?? [];

		a.push(item);

		map.set(id, a);

		return map
	}, init?.() ?? new Map() as Map<any, T[]>);
}

export function sumGroup<T extends Record<any, any[]> | Map<any, any[]>>(group: T)
{
	let len = 0;

	if (typeof group.forEach === 'undefined')
	{
		// @ts-ignore
		group = Object.values(group);
	}

	// @ts-ignore
	group.forEach(b => len += b.length);

	return len;
}

export default arrayGroupToRecord
