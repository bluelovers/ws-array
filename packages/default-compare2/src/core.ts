import { _typeOf } from './util';

export const enum EnumSortCompareResult
{
	KEEP = 0,
	RIGHT = 1,
	LEFT = -1,
}

export function numberCompare(a: unknown, b: unknown)
{
	if (a < b)
	{
		return EnumSortCompareResult.LEFT
	}
	else if (a > b)
	{
		return EnumSortCompareResult.RIGHT
	}

	return EnumSortCompareResult.KEEP
}

/**
 * will return undefined when a and b not null/undefined
 */
export function nullCompare(a: unknown, b: unknown)
{
	const typeA = _typeOf(a);
	const typeB = _typeOf(b);

	if (typeA === 'null')
	{
		return typeB === 'null' ? EnumSortCompareResult.KEEP : (typeB === 'undefined'
			? EnumSortCompareResult.LEFT
			: EnumSortCompareResult.RIGHT);
	}
	else if (typeA === 'undefined')
	{
		return typeB === 'null' ? EnumSortCompareResult.RIGHT : (typeB === 'undefined'
			? EnumSortCompareResult.KEEP
			: EnumSortCompareResult.RIGHT);
	}
	else if (typeB === 'null' || typeB === 'undefined')
	{
		return EnumSortCompareResult.LEFT;
	}
}

export function symbolCompare(a: unknown, b: unknown)
{
	const typeA = _typeOf(a);
	const typeB = _typeOf(b);

	if (typeA === 'symbol')
	{
		return typeB === 'symbol' ? numberCompare(String(a), String(b)) : EnumSortCompareResult.LEFT
	}
	else if (typeB === 'symbol')
	{
		return EnumSortCompareResult.RIGHT
	}

	return numberCompare(a, b)
}

export function dateCompare(a: Date, b: Date)
{
	return a.getTime() - b.getTime()
}

export function stringCompareLocale(a: string, b: string)
{
	return a.localeCompare(b)
}
