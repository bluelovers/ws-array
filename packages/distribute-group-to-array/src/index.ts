export interface IOptionsDistributeGroupToArray<T>
{
	init?(): T[][],
	groupArraySize?: number,
}

export function _createGroupArray<T, N extends number = 4>(length: N): T[][] & {
	readonly length: N
}
{
	const arr: T[][] = [];

	for (let i = 0; i < length; i++)
	{
		arr.push([]);
	}

	return arr as any;
}

export function distributeGroupToArrayCore<T>(group: Record<string, T[]> | Map<any, T[]>,
	options?: IOptionsDistributeGroupToArray<T>,
)
{
	// @ts-ignore
	if (typeof group.entries !== 'undefined')
	{
		group = mapToRecord(group as any);
	}

	// @ts-ignore
	const arr = options?.init?.() ?? _createGroupArray<T>(options?.groupArraySize ?? 4);

	const len = arr.length;

	const { names, record } = Object.entries(group).reduce((a, [k, v]) =>
	{

		a.names.push(k);
		a.record[k] = v.slice();

		return a
	}, {
		names: [] as string[],
		record: {} as Record<string, T[]>,
	});

	let i = len - 1;

	do
	{
		let item: T;

		for (let name of names)
		{
			item = record[name]?.pop();

			if (item)
			{
				arr[i % len].push(item);
			}
			else
			{
				delete record[name]
			}
		}

		if (item)
		{
			i++;
		}

	}
	while (Object.keys(record).length)

	return arr
}

export function mapToRecord<T>(map: Map<any, T[]>)
{
	const record = {} as Record<string, T[]>;

	for (let [k, v] of map.entries())
	{
		record[k] = v;
	}

	return record
}

export function distributeGroupToArray<T>(group: Record<string, T[]> | Map<any, T[]>,
	options?: IOptionsDistributeGroupToArray<T>,
)
{
	return distributeGroupToArrayCore<T>(group as any, options).flat()
}

export default distributeGroupToArray
