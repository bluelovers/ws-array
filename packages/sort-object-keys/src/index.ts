
export function sortObject<T extends Record<any, any>>(object: T, sortFn: IOptions<T>["sort"]): T
export function sortObject<T extends Record<any, any>>(object: T, keyOrders: IOptions<T>["keys"]): T
export function sortObject<T extends Record<any, any>>(object: T, options?: Omit<IOptions<T>, 'keys' | 'onlyKeys'> & {
	keys: IOptions<T>["keys"],
	onlyKeys: true,
}): Partial<T>
export function sortObject<T extends Record<any, any>>(object: T, options?: Omit<IOptions<T>, 'useSource'> & {
	useSource: true,
}): T
export function sortObject<T extends Record<any, any>>(object: T, options?: IOptions<T>): T
export function sortObject<T extends Record<any, any>>(object: T, sortWith): T | Partial<T>
{
	let options: IOptions<T> = {};

	if (typeof sortWith === 'function')
	{
		options.sort = sortWith;
	}
	else if (Array.isArray(sortWith))
	{
		options.keys = sortWith;
	}
	else
	{
		options = Object.assign(options, sortWith);
	}

	let {
		keys = [] as IOptions<T>["keys"],
		useSource,
	} = options;

	if (options.onlyKeys)
	{
		useSource = false;

		if (typeof keys.length !== 'number' || keys.length === 0)
		{
			throw new ReferenceError(`options.key is empty or not exists.`)
		}
	}
	else
	{
		keys = keys.concat()
			.concat(Object.keys(object).sort(options.sort))
		;
	}

	keys = array_unique(keys);

	if (options.desc)
	{
		keys = keys.reverse()
	}

	let ret = keys.reduce(function (total, key)
	{
		if (options.allowNotExists || key in object)
		{
			total[key] = object[key];
		}

		return total;
	}, {} as Partial<T>);

	if (useSource)
	{
		(Object.keys(ret) as IOptions<T>["keys"])
			.forEach(function (key, index, array)
			{
				delete object[key];
				object[key] = ret[key];
			})
		;

		return object;
	}

	return ret;
}

export interface IOptions<T extends Record<any, any> = Record<any, any> , K extends string = Extract<keyof T, string>>
{
	/**
	 * key order
	 */
	keys?: (string | K)[],
	/**
	 * return Object only keys
	 * will disable useSource
	 */
	onlyKeys?: boolean,
	/**
	 * sort callback
	 *
	 * @param a
	 * @param b
	 * @returns {any}
	 */
	sort?: (a: string | K, b: string | K) => number,
	/**
	 * return reversed Object
	 */
	desc?: boolean,
	allowNotExists?: boolean,
	/**
	 * return source Object
	 */
	useSource?: boolean,
}

sortObject.sortObjectKeys = sortObject;

// @ts-ignore
(sortObject as any).default = sortObject;

export { sortObject as sortObjectKeys }

function array_unique(array: any[])
{
	return array.filter(function (el, index, arr)
	{
		return index === arr.indexOf(el);
	});
}

export default sortObject;
