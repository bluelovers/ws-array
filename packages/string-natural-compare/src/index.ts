/**
 * Created by user on 2020/6/4.
 */

// @ts-ignore
import _naturalCompare from 'string-natural-compare';

export interface IOptionsStringNaturalCompare
{
	/**
	 * Set to true to compare strings case-insensitively. Default: false.
	 */
	caseInsensitive?: boolean,
	/**
	 * A string of characters that define a custom character ordering. Default: undefined.
	 */
	alphabet?: string,
}

export interface IOptionsNaturalCompare extends IOptionsStringNaturalCompare
{
	desc?: boolean,
}

/**
 * Compare alphanumeric strings the same way a human would,
 * using a natural order algorithm
 * (originally known as the alphanum algorithm)
 * where numeric characters are sorted
 * based on their numeric values rather than their ASCII values.
 */
export function naturalCompare(a: string | number, b: string | number, opts?: IOptionsNaturalCompare)
{
	let i: number;

	if (typeof a === 'number' && typeof b === 'number')
	{
		i = a - b
	}
	else
	{
		if (typeof a === 'number')
		{
			a = String(a);
		}
		else if (typeof b === 'number')
		{
			b = String(b);
		}

		if (a === b)
		{
			return 0
		}

		i = _naturalCompare(a, b, opts)
	}

	if (i !== 0 && opts?.desc)
	{
		i = 0 - i;
	}

	return i
}

/**
 * create compare with preset options
 */
export function createNew(opts?: IOptionsNaturalCompare)
{
	return (a: string | number, b: string | number) => naturalCompare(a, b, opts)
}

/**
 * compare strings case-insensitively
 */
export const compareCaseInsensitive = createNew({
	caseInsensitive: true,
})

export { compareCaseInsensitive as caseInsensitive }

naturalCompare.createNew = createNew;
naturalCompare.compareCaseInsensitive = compareCaseInsensitive;
naturalCompare.caseInsensitive = compareCaseInsensitive;
naturalCompare.default = naturalCompare;

Object.defineProperty(naturalCompare, "__esModule", { value: true });

export default naturalCompare
