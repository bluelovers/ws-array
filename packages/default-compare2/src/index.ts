import { nullCompare, numberCompare, symbolCompare } from './core';

export * from './util';
export * from './core';

export function defaultCompareBasic(a: unknown, b: unknown)
{
	const result = nullCompare(a, b);

	if (typeof result === 'number')
	{
		return result
	}

	return numberCompare(a, b)
}

export function defaultCompareWithSymbol(a: unknown, b: unknown)
{
	const result = nullCompare(a, b);

	if (typeof result === 'number')
	{
		return result
	}

	return symbolCompare(a, b)
}

export function arraySortWithSymbol<T extends any[]>(arr: T)
{
	return arr.sort(defaultCompareWithSymbol)
}

export { defaultCompareWithSymbol as defaultCompare }

export default defaultCompareWithSymbol
