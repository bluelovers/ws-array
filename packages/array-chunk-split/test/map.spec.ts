import { arrayChunkMap } from '../src/index';

describe(`arrayChunkMap`, () =>
{

	test(`default`, () =>
	{

		const arr = [1, 2, 3, 4, 5, 6, 7, 8];

		let actual = arrayChunkMap({
			inputArray: arr,
			maxChunkLength: 2,
		});
		let expected = [ 1, 5 ];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`mapMethod:true`, () =>
	{

		const arr = [1, 2, 3, 4, 5, 6, 7, 8];

		let actual = arrayChunkMap({
			inputArray: arr,
			maxChunkLength: 2,
			mapMethod: true,
		});
		let expected = [ 4, 8 ];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`maxChunkSize`, () =>
	{

		const arr = [1, 2, 3, 4, 5, 6, 7, 8];

		let actual = arrayChunkMap({
			inputArray: arr,

			mapMethod(value, index)
			{
				return index
			},

			maxChunkSize: 2,
		});
		let expected = [ 0, 1, 2, 3 ];

		expect(actual).toStrictEqual(expected);
		//expect(actual).toBeInstanceOf(Date);
		expect(actual).toMatchSnapshot();

	});

	test(`throw error`, () =>
	{
		expect(() => arrayChunkMap({
			inputArray: [],
		} as any)).toThrowError();
	})

})
