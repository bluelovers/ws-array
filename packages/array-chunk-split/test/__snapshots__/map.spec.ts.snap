// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arrayChunkMap default 1`] = `
Array [
  1,
  5,
]
`;

exports[`arrayChunkMap mapMethod:true 1`] = `
Array [
  4,
  8,
]
`;

exports[`arrayChunkMap maxChunkSize 1`] = `
Array [
  0,
  1,
  2,
  3,
]
`;
